--- gpio/Makefile	2016-01-15 17:22:35.878983341 +0100
+++ gpio/Makefile	2016-01-15 17:21:39.733695350 +0100
@@ -65,11 +65,12 @@ tags:	$(SRC)
 .PHONY:	install
 install:
 	@echo "[Install]"
-	@cp gpio		$(DESTDIR)$(PREFIX)/bin
+	@install -m 0755 -d	${DESTDIR}${PREFIX}/bin
+	@cp gpio		$(DESTDIR)$(PREFIX)/bin/gpio
 	@chown root.root	$(DESTDIR)$(PREFIX)/bin/gpio
 	@chmod 4755		$(DESTDIR)$(PREFIX)/bin/gpio
-	@mkdir -p		$(DESTDIR)$(PREFIX)/man/man1
-	@cp gpio.1		$(DESTDIR)$(PREFIX)/man/man1
+	@mkdir -p		$(DESTDIR)$(PREFIX)/share/man/man1
+	@cp gpio.1		$(DESTDIR)$(PREFIX)/share/man/man1
 
 .PHONY:	uninstall
 uninstall:

