diff -ur fbida-2.12/gfx.h fbida-2.12/gfx.h
--- fbida-2.12/gfx.h	2016-09-07 19:48:07.719421934 +0200
+++ fbida-2.12/gfx.h	2016-09-07 19:48:25.999421927 +0200
@@ -1,8 +1,6 @@
 #include <stdbool.h>
 #include <inttypes.h>
 
-#include <epoxy/egl.h>
-
 typedef struct gfxstate gfxstate;
 
 struct gfxstate {
@@ -17,11 +15,6 @@
     uint32_t rlen, glen, blen, tlen;
     uint32_t roff, goff, boff, toff;
 
-    /* egl */
-    EGLDisplay dpy;
-    EGLContext ctx;
-    EGLSurface surface;
-
     /* calls */
     void (*restore_display)(void);
     void (*cleanup_display)(void);

